### DISCUSSION TITLE
Generics that extends an interface

### DISCUSSION THREAD

**Discussion by StunxFS** [vlang/v#11723] [Ideas]
Rust, if you have programmed in it, allows you to specify a generic to a specific trait. I don't remember what this feature is called, but it looks like this:
```rust
trait Operable {
    fn operate() -> String;
}

fn gen<T: Operable>(operator: T) {
    println!("{}", operator.operate());
}
```
The purpose of this is to make sure that the type that will be passed to the function fulfills a certain function or field, to avoid making use of non-existent things.

In V this could be elegant, because with this we could tell the compiler that we expect a type that implements one (or more) specific interfaces.
```v
interface Operable {
    operate() string
}

interface Hack {
    hackme()
}

fn gen<T: Operable + Hack>(v T) {
    println(v.operate())
    v.hackme()
}
```
What do you think about this?

**Comment by danieldaeschle** [comment]
yup, that will be implemented.
might look like:
```v
interface Operable {
    operate() string
}

interface Hack {
    hackme()
}

fn gen<T = Operable + Hack>(v T) {  // or even extends?
    println(v.operate())
    v.hackme()
}
```

@medvednikov has to decide.

**Comment by smartiniOnGitHub** [comment]
Hi all, great idea :-) .
What do you think only to align to sum types syntax (using the "|" to identify all interfaces allowed) ?
I think in TypeScript there are a lot of interesting ideas about this (and generics related things: constraints, etc) and a common syntax, see [here](https://www.typescriptlang.org/docs/handbook/generics.html) ... hope it's a useful suggestion. Bye

**Comment by ntrel** [comment]
> using the "|" to identify all interfaces allowed) ?

For sum types that means 'one of these types'. Here we want *all* of the interfaces to be satisfied.

Go [proposes](https://go.googlesource.com/proposal/+/refs/heads/master/design/go2draft-type-parameters.md#using-a-constraint) to use no punctuation: `GenericType InterfaceName`:
```go
type Stringer interface {
	String() string
}
func Stringify[T Stringer](s []T) (ret []string) {
```
I don't think they support listing more than one constraint for the same type, but we could allow them inline, e.g. `T Operable Hack`.

BTW I don't think we need to support type list constraints:
```go
type Float interface {
	type float32, float64
}
```
We have sum types so we could use `type Float = f32 | f64` if we want to support this for a type constraint.

**Comment by ntrel** [comment]
Operator overloads could be included in an interface. In that case basic types like `int` could implement it so long as it satisfies all the operator methods in the interface.

```v
interface Arithmetic<T> {
  +(right T) T
  -(right T) T
}

fn foo<T Arithmetic<T>>()

foo<int>() // instantiates `Arithmetic<int>`. OK because `int` supports `+`, `-`.
foo<Bar>() // instantiates `Arithmetic<Bar>`. OK if Bar overloads `+`, `-`.
```

**Comment by mcastorina** [comment]
Offering an alternative idea: Go solves this with interface embeds. It would look like this in V:

```v
interface Operable {
    operate() string
}

interface Hack {
    hackme()
}

interface OperableHack {
    Operable
    Hack
}

fn gen(v OperableHack) {
    println(v.operate())
    v.hackme()
}
```

It could even be inlined:
```v
fn gen(v interface{ Operable Hack }) {
   // ...
}
```

I'm not sure I understand why interfaces and generics are being mixed in this case.
